test/source/coiteration_test.cpp

```diff
@@ -298,3 +308,122 @@ TEST_CASE("Coiteration-Dense-Hashed-ConjunctiveMerge")
+    // respectively
+    xsparse::level_capabilities::Coiterate<
+        xsparse::util::LambdaWrapper<decltype(fn)>::template apply,
+        decltype(fn),
+        uintptr_t,
+        uintptr_t,
+        std::tuple<decltype(d1), decltype(d2)>,
+        std::tuple<std::tuple<>>,
+        std::tuple<uintptr_t, uintptr_t>>
+        coiter_dense(fn, d1, d2);
+
+    xsparse::level_capabilities::Coiterate<
+        xsparse::util::LambdaWrapper<decltype(fn)>::template apply,
+        decltype(fn),
+        uintptr_t,  // IK
+        uintptr_t,  // PK
+        std::tuple<decltype(s1), decltype(s2)>,
+        std::tuple<std::tuple<uintptr_t>>,  // tuple of Is, which is passed to coiteration_helper
+        std::tuple<std::optional<uintptr_t>, std::optional<uintptr_t>>>
+        coiter_compressed(fn, s1, s2);
+
+    auto it_helper1 = d1.iter_helper(std::make_tuple(), ZERO);
+    auto it1 = it_helper1.begin();
+    auto it_helper2 = d2.iter_helper(std::make_tuple(), ZERO);
+    auto it2 = it_helper2.begin();
+
+    auto end1 = it_helper1.end();
+    auto end2 = it_helper2.end();
+
+    // when co-iterating over levels that are unordered (i.e. hashed), then we use locate to
+    // check if the index exists in the hashed level. If not, then we skip it.
+    for (auto const [ik, pk_tuple] : coiter_dense.coiter_helper(std::make_tuple(std::make_tuple()),
+                                                                std::make_tuple(ZERO, ZERO)))
+    {
+        // get the index and pointer from the outer-most level involved in co-iteration
+        auto [i1, p1] = *it1;
+        auto [i2, p2] = *it2;
+
+        // the index should be the same for both levels and tracked via the co-iterator
+        CHECK(ik == i1);
+        CHECK(ik == i2);
+        CHECK(std::get<0>(pk_tuple).value() == p1);
+        CHECK(std::get<1>(pk_tuple).value() == p2);
+
+        // use these to define the inner-most iterator
+        // auto it_helper_inner1 = s1.iter_helper(i1, p1);
+        // auto it_helper_inner2 = s2.iter_helper(i2, p2);
+        // auto it1_inner = it_helper_inner1.begin();
+        // auto it2_inner = it_helper_inner2.begin();
+        // auto end1_inner = it_helper_inner1.end();
+        // auto end2_inner = it_helper_inner2.end();
+
+        // co-iterate over the inner-most compressed level now
+        // for (auto const [cik, cpk_tuple] :
+        //      coiter_compressed.coiter_helper(std::make_tuple(std::make_tuple(ik)), pk_tuple))
+        // {
+        //     auto [ci1, cp1] = *it1_inner;
+        //     auto [ci2, cp2] = *it2_inner;
+        //     uintptr_t l = std::min(ci1, ci2);
+
+        //     if (ci1 == l)
+        //     {
+        //         CHECK(cp1 == std::get<0>(cpk_tuple).value());
+        //         ++it1_inner;
+        //     }
+        //     if (ci2 == l)
+        //     {
+        //         CHECK(cp2 == std::get<1>(cpk_tuple).value());
+        //         ++it2_inner;
+        //     }
+        // }
```

Comment:
When these LOC get uncommented, there is a segfault error during compilation time that I cannot parse.

```
(base) adam2392@Adams-MBP-3 xsparse % cmake --build build/test
[ 25%] Built target fmt
[ 25%] Built target XSparse
[ 33%] Building CXX object CMakeFiles/XSparseTests.dir/source/coiteration_test.cpp.o
clang: error: unable to execute command: Segmentation fault: 11
clang: error: clang frontend command failed due to signal (use -v to see invocation)
Apple clang version 14.0.3 (clang-1403.0.22.14.1)
Target: arm64-apple-darwin22.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
clang: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /var/folders/6_/sl83qtkd68x3_mvfys07_6qm0000gn/T/coiteration_test-d19547.cpp
clang: note: diagnostic msg: /var/folders/6_/sl83qtkd68x3_mvfys07_6qm0000gn/T/coiteration_test-d19547.sh
clang: note: diagnostic msg: Crash backtrace is located in
clang: note: diagnostic msg: /Users/adam2392/Library/Logs/DiagnosticReports/clang_<YYYY-MM-DD-HHMMSS>_<hostname>.crash
clang: note: diagnostic msg: (choose the .crash file that corresponds to your crash)
clang: note: diagnostic msg: 

********************
make[2]: *** [CMakeFiles/XSparseTests.dir/source/coiteration_test.cpp.o] Error 254
make[1]: *** [CMakeFiles/XSparseTests.dir/all] Error 2
make: *** [all] Error 2
```

Reply:
This seems to be a bug on clang/Apple's side.

Reply:
I'd debug on CI, or install g++ using homebrew and use that instead, or use a Linux VM.

Reply:
Ah okay that helped. It seems the issue is that when we dereference `coiterate`, it returns a `std::optional<long unsigned int>`, but the iter_helper is expecting `BaseTraits::PKM1` type, which is defined as 

```
using PKM1 = std::tuple_element_t<
            // First element of P
            0,
            // Except that the type list has uint8_t appended to it.
            // This has the effect of choosing uint8_t instead of the
            // first element if `sizeof...(LowerLevels) == 0`.
            // is in place.
            std::tuple<position_t<LowerLevels>..., bool>>;
```

Complete error message:
```
In file included from /Users/adam2392/Documents/xsparse/test/source/coiteration_test.cpp:19:
/Users/adam2392/Documents/xsparse/include/xsparse/level_capabilities/co_iteration.hpp: In instantiation of 'auto unfold_and_apply_helper(const std::tuple<_UTypes ...>&, const ITuple&, const Pkm1Tuple&, std::index_sequence<Indices ...>) [with Args = {xsparse::levels::compressed<std::tuple<xsparse::levels::dense<std::tuple<>, long unsigned int, long unsigned int, xsparse::level_properties<true, true, true, false, true> > >, long unsigned int, long unsigned int, xsparse::util::container_traits<std::vector, std::set, std::map>, xsparse::level_properties<true, true, false, false, true> >&, xsparse::levels::compressed<std::tuple<xsparse::levels::dense<std::tuple<>, long unsigned int, long unsigned int, xsparse::level_properties<true, true, true, false, true> > >, long unsigned int, long unsigned int, xsparse::util::container_traits<std::vector, std::set, std::map>, xsparse::level_properties<true, true, false, false, true> >&}; ITuple = std::tuple<long unsigned int>; Pkm1Tuple = std::tuple<std::optional<long unsigned int>, std::optional<long unsigned int> >; long unsigned int ...Indices = {0, 1}; std::index_sequence<Indices ...> = std::integer_sequence<long unsigned int, 0, 1>]':
/Users/adam2392/Documents/xsparse/include/xsparse/level_capabilities/co_iteration.hpp:188:56:   required from 'xsparse::level_capabilities::Coiterate<F, Ffunc, IK, PK, std::tuple<Levels ...>, std::tuple<Is ...>, std::tuple<Ps ...> >::coiteration_helper::coiteration_helper(const xsparse::level_capabilities::Coiterate<F, Ffunc, IK, PK, std::tuple<Levels ...>, std::tuple<Is ...>, std::tuple<Ps ...> >&, std::tuple<Is ...>, std::tuple<Ps ...>) [with F = xsparse::util::LambdaWrapper<DOCTEST_ANON_FUNC_10()::<lambda(std::tuple<bool, bool>)> >::apply; Ffunc = DOCTEST_ANON_FUNC_10()::<lambda(std::tuple<bool, bool>)>; IK = long unsigned int; PK = long unsigned int; Levels = {xsparse::levels::compressed<std::tuple<xsparse::levels::dense<std::tuple<>, long unsigned int, long unsigned int, xsparse::level_properties<true, true, true, false, true> > >, long unsigned int, long unsigned int, xsparse::util::container_traits<std::vector, std::set, std::map>, xsparse::level_properties<true, true, false, false, true> >, xsparse::levels::compressed<std::tuple<xsparse::levels::dense<std::tuple<>, long unsigned int, long unsigned int, xsparse::level_properties<true, true, true, false, true> > >, long unsigned int, long unsigned int, xsparse::util::container_traits<std::vector, std::set, std::map>, xsparse::level_properties<true, true, false, false, true> >}; Is = {long unsigned int}; Ps = {std::optional<long unsigned int>, std::optional<long unsigned int>}]'
/Users/adam2392/Documents/xsparse/include/xsparse/level_capabilities/co_iteration.hpp:441:55:   required from 'xsparse::level_capabilities::Coiterate<F, Ffunc, IK, PK, std::tuple<Levels ...>, std::tuple<Is ...>, std::tuple<Ps ...> >::coiteration_helper xsparse::level_capabilities::Coiterate<F, Ffunc, IK, PK, std::tuple<Levels ...>, std::tuple<Is ...>, std::tuple<Ps ...> >::coiter_helper(std::tuple<Is ...>, std::tuple<Ps ...>) [with F = xsparse::util::LambdaWrapper<DOCTEST_ANON_FUNC_10()::<lambda(std::tuple<bool, bool>)> >::apply; Ffunc = DOCTEST_ANON_FUNC_10()::<lambda(std::tuple<bool, bool>)>; IK = long unsigned int; PK = long unsigned int; Levels = {xsparse::levels::compressed<std::tuple<xsparse::levels::dense<std::tuple<>, long unsigned int, long unsigned int, xsparse::level_properties<true, true, true, false, true> > >, long unsigned int, long unsigned int, xsparse::util::container_traits<std::vector, std::set, std::map>, xsparse::level_properties<true, true, false, false, true> >, xsparse::levels::compressed<std::tuple<xsparse::levels::dense<std::tuple<>, long unsigned int, long unsigned int, xsparse::level_properties<true, true, true, false, true> > >, long unsigned int, long unsigned int, xsparse::util::container_traits<std::vector, std::set, std::map>, xsparse::level_properties<true, true, false, false, true> >}; Is = {long unsigned int}; Ps = {std::optional<long unsigned int>, std::optional<long unsigned int>}]'
/Users/adam2392/Documents/xsparse/test/source/coiteration_test.cpp:404:45:   required from here
/Users/adam2392/Documents/xsparse/include/xsparse/level_capabilities/co_iteration.hpp:45:63: error: cannot convert 'std::__tuple_element_t<0, std::tuple<std::optional<long unsigned int>, std::optional<long unsigned int> > >' {aka 'const std::optional<long unsigned int>'} to 'xsparse::util::base_traits<xsparse::levels::compressed, std::tuple<xsparse::levels::dense<std::tuple<>, long unsigned int, long unsigned int, xsparse::level_properties<true, true, true, false, true> > >, long unsigned int, long unsigned int, xsparse::util::container_traits<std::vector, std::set, std::map>, xsparse::level_properties<true, true, false, false, true> >::PKM1' {aka 'long unsigned int'}
   45 |     return std::make_tuple(std::get<Indices>(args).iter_helper(m_i, std::get<Indices>(pkm1))...);
      |                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /Users/adam2392/Documents/xsparse/include/xsparse/levels/compressed.hpp:10,
                 from /Users/adam2392/Documents/xsparse/test/source/coiteration_test.cpp:11:
/Users/adam2392/Documents/xsparse/include/xsparse/level_capabilities/coordinate_iterate.hpp:271:90: note:   initializing argument 2 of 'xsparse::level_capabilities::coordinate_position_iterate<T, IK, PK, LowerLevels>::iteration_helper xsparse::level_capabilities::coordinate_position_iterate<T, IK, PK, LowerLevels>::iter_helper(typename BaseTraits::I, typename BaseTraits::PKM1) [with T = xsparse::levels::compressed; IK = std::tuple<xsparse::levels::dense<std::tuple<>, long unsigned int, long unsigned int, xsparse::level_properties<true, true, true, false, true> > >; PK = long unsigned int; LowerLevels = {long unsigned int, xsparse::util::container_traits<std::vector, std::set, std::map>, xsparse::level_properties<true, true, false, false, true>}; typename BaseTraits::I = std::tuple<long unsigned int>; BaseTraits = xsparse::util::base_traits<xsparse::levels::compressed, std::tuple<xsparse::levels::dense<std::tuple<>, long unsigned int, long unsigned int, xsparse::level_properties<true, true, true, false, true> > >, long unsigned int, long unsigned int, xsparse::util::container_traits<std::vector, std::set, std::map>, xsparse::level_properties<true, true, false, false, true> >; typename BaseTraits::PKM1 = long unsigned int]'
  271 |         iteration_helper iter_helper(typename BaseTraits::I i, typename BaseTraits::PKM1 pkm1)
      |                                                                ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
```

Reply:
It's just wrapped in an extra `optional`, I think it should be safe to get the value.

Reply:
Possibly stupid question. Where would you suggest making the change then? Perhaps in 

```
template <typename... Args, typename ITuple, typename Pkm1Tuple, size_t... Indices>
auto
unfold_and_apply_helper(const std::tuple<Args...>& args,
                        const ITuple& m_i,
                        const Pkm1Tuple& pkm1,
                        std::index_sequence<Indices...>)
{
    return std::make_tuple(std::get<Indices>(args).iter_helper(m_i, std::get<Indices>(pkm1))...);
}
```
? But not every element of `pkm1` is `std::optional`

Reply:
I think `pkm1` should never be `optional` in the first place, something is going wrong up the chain.

Reply:
The issue is that dereferencing either calls: `deref_PKs(i)`, or `std::get<I>(this->m_coiterHelper.m_coiterate.m_levelsTuple)
                            .locate(std::get<I>(m_coiterHelper.m_pkm1), min_ik)` depending on whether the level is ordered or not.

Deref_PKS:

```
template <class iter>
                inline auto deref_PKs(iter i) const noexcept
                {
                    return (std::get<0>(*i) == min_ik);
                               ? std::optional<std::tuple_element_t<1, decltype(*i)>>(
                                   std::get<1>(*i))
                               : std::nullopt;
                }
```

Similarly locate is defined as:

```
inline auto locate(typename BaseTraits::PKM1 pkm1, IK ik) const noexcept
            {
                auto it = m_crd[pkm1].find(ik);
                return it != m_crd[pkm1].end() ? std::optional<PK>(it->second) : std::nullopt;
            }
```

So both routes return std::optional.

Reply:
If the algorithm is written correctly we should be able to de reference both without issue.

Reply:
We can dereference both though as evidenced by the existing unit-tests that test both of these dereferences of the `Coiterate`. That is not the issue I believe. The issue is that the compiler cannot handle the following:

```
'std::__tuple_element_t<0, std::tuple<std::optional<long unsigned int>, std::optional<long unsigned int> > >' {aka 'const std::optional<long unsigned int>'} to 

'xsparse::util::base_traits<xsparse::levels::compressed, std::tuple<xsparse::levels::dense<std::tuple<>, long unsigned int, long unsigned int, xsparse::level_properties<true, true, true, false, true> > >, long unsigned int, long unsigned int, xsparse::util::container_traits<std::vector, std::set, std::map>, xsparse::level_properties<true, true, false, false, true> >::PKM1' {aka 'long unsigned int'}
```

Reply:
Ah I now see the issue -- we are missing const in one of them.

Reply:
Google const correctness; something should be const but isn't or the other way around.

Reply:
Ah okay it compiles! I needed to change one of the template arguments defined the `Ps` template type in `Coiterate`. It was incorrectly adding a `std::optional`, which I think gave it an issue because the returned type during dereferencing is "there".

However, now the unit-test is the main issue. It is possible there is either an error in my unit-test, or in the actual Coiterate implementation for a nested level.